src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs

```diff
@@ -74,7 +74,17 @@ public virtual async Task StopAsync(CancellationToken cancellationToken)
             finally
             {
                 // Wait until the task completes or the stop token triggers
-                await Task.WhenAny(_executeTask, Task.Delay(Timeout.Infinite, cancellationToken)).ConfigureAwait(false);
+                try
```

Comment:
@stephentoub - do you have any opinion on this approach? The options I see are:

1. The current code.
2. Explicitly not awaiting the Task returned from awaiting `Task.WhenAny`, which will ignore the OperationCanceledException - because we never await the Task that got canceled.
```C#
var tcs = new TaskCompletionSource<object>();
using CancellationTokenRegistration registration = cancellationToken.Register(s => ((TaskCompletionSource<object>)s).SetCanceled(), tcs);
await Task.WhenAny(_executeTask, tcs.Task).ConfigureAwait(false);
```

I'm not 100% sure which is the best option. I thought the current proposed approach was more explicit as to what is happening: An OperationCanceledException is being thrown, and we are explicitly ignoring it.

Performance here isn't a major concern because these BackgroundServices don't get Start/Stopped very often (typically just at the beginning and end of the process), so I didn't think the exception handling would make a perf difference.

Reply:
I spoke with @stephentoub offline. His recommendation was to keep it simple, and just use option (2) above on all TFMs. I also recommend adding a comment that we explicitly aren't `await`ing the `_executeTask` here, just the `Task.WhenAny` because cancelling the `_executeTask` will always throw an `OperationCanceledException` and we are just ignoring that exception.

Futher, @stephentoub showed me that in "normal" hosting scenarios, we don't actually have a leak with the current code. The Task created from `Task.Delay(Timeout.Infinite, cancellationToken)` will be available for GC once the `cancellationToken` passed in goes away. This is because `Timeout.Infinite` is special cased to not create a Timer:

https://github.com/dotnet/runtime/blob/62948580b4d3547473cb567bff6cc61308136d81/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L5619-L5629

Instead, the CancellationToken and the Task from `Task.WhenAny` are the things holding onto the Task. The latter will be available for GC once this method completes. So the only question is the lifetime of the CancellationToken.

In normal hosting scenarios, `StopAsync` gets called with a CancellationToken from a CancellationTokenSource that is disposed of after all the hosted services are stopped:

https://github.com/dotnet/runtime/blob/62948580b4d3547473cb567bff6cc61308136d81/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L122-L135

In that case, there won't be a leak.

However, since this API is public, taking a fix here is the right thing to do since anyone can call this method with any CancellationToken - for example one that is long-lived.